File: apps/souls-studio/src/types/soul.ts
export interface Soul {
  id: string
  name: string
  description: string
  owner: string
  repo: string
  soulsPath: string
  path: string
  content?: string
  isInstalled: boolean
  isFetched: boolean
}

export interface SoulRepository {
  owner: string
  repo: string
  soulsPath: string
  lastFetched?: string
}

export interface Catalog {
  version: string
  lastUpdated: string
  souls: CatalogSoul[]
}

export interface CatalogSoul {
  id: string
  name: string
  description: string
  owner: string
  repo: string
  soulsPath: string
  path: string
}

export interface FetchedRepos {
  repos: Record<string, string>
}

export type InstallMethod = 'npx' | 'copy'
export type ThemeMode = 'system' | 'light' | 'dark'

export type OpenClawConnectionMethod = 'local' | 'ssh' | 'tailscale' | 'direct'

export interface OpenClawConnection {
  id: string
  name: string
  method: OpenClawConnectionMethod
  gatewayUrl: string
  authToken?: string
  sshHost?: string
  sshUser?: string
  sshPort?: number
  sshKeyPath?: string
  sshPassphrase?: string
  sshPassword?: string
  isDefault?: boolean
  discovered?: boolean
}

export interface GatewayAgent {
  agentId: string
  hasSoulMd: boolean
  hasAgentsMd: boolean
  hasToolsMd: boolean
}

export type GatewayConnectionStatus = 'disconnected' | 'connecting' | 'connected' | 'error'

export type AgentFileType = 'soul.md' | 'tools.md' | 'agents.md'

export interface SshHostCandidate {
  hostAlias: string
  hostname: string
  user?: string
  port: number
  identityFile?: string
  gatewayResponds: boolean
}

export interface Settings {
  installMethod: InstallMethod
  theme: ThemeMode
  editorFontSize: number
  syncIntervalSeconds: number
  openclawConnections: OpenClawConnection[]
}

export interface RepoGroup {
  owner: string
  repo: string
  souls: Soul[]
  isFetched: boolean
  isCustom?: boolean
}

export interface RepoInfo {
  owner: string
  repo: string
  isFetched: boolean
  isCustom: boolean
  highlight: boolean
}

export type Selection =
  | { type: 'none' }
  | { type: 'repo'; repo: RepoGroup }
  | { type: 'soul'; soul: Soul }

export interface Favorites {
  souls: string[]
  repos: string[]
}

export interface SoulSnapshot {
  name: string
  slug: string
  tagline: string
  description: string
  content: string
  updatedAt: number
}

export type LocalSoulSyncState = 'synced' | 'pending' | 'error'

export interface LocalSoul {
  localId: string
  remoteId?: string
  ownerHandle?: string
  slug: string
  name: string
  tagline: string
  description: string
  content: string
  localUpdatedAt: number
  remoteUpdatedAt?: number
  syncState: LocalSoulSyncState
  snapshot?: SoulSnapshot
  deleted?: boolean
  lastError?: string
}

export type SyncOperationType = 'publish' | 'updateMetadata' | 'deleteSoul'

export interface SyncOperation {
  id: string
  localSoulId: string
  type: SyncOperationType
  createdAt: number
  retryCount: number
  lastError?: string
}

export interface SoulsStoreMeta {
  schemaVersion: number
  lastPullAt?: number
  lastPushAt?: number
}

export interface SoulsStore {
  records: LocalSoul[]
  ops: SyncOperation[]
  meta: SoulsStoreMeta
}

export type SyncState = 'idle' | 'syncing' | 'offline' | 'error'

export interface SyncStatus {
  state: SyncState
  pendingCount: number
  lastSyncAt?: number
  cappedByLimit: boolean
  lastError?: string
}

export const DEFAULT_SOULS_STORE: SoulsStore = {
  records: [],
  ops: [],
  meta: {
    schemaVersion: 1,
  },
}

File: apps/souls-studio/src-tauri/src/models/soul.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Soul {
    pub id: String,
    pub name: String,
    pub description: String,
    pub owner: String,
    pub repo: String,
    pub souls_path: String,
    pub path: String,
    pub content: Option<String>,
    pub is_installed: bool,
    pub is_fetched: bool,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Catalog {
    pub version: String,
    pub last_updated: String,
    pub repos: Vec<CatalogRepo>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CatalogRepo {
    pub url: String,
    #[serde(default)]
    pub highlight: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct FetchedRepos {
    pub repos: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct OpenClawConnection {
    pub id: String,
    pub name: String,
    pub method: String,
    pub gateway_url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_token: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_host: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_user: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_port: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_key_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_passphrase: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssh_password: Option<String>,
    #[serde(default)]
    pub is_default: bool,
    #[serde(default)]
    pub discovered: bool,
}

#[allow(dead_code)]
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GatewayAgent {
    pub agent_id: String,
    #[serde(default)]
    pub has_soul_md: bool,
    #[serde(default)]
    pub has_agents_md: bool,
    #[serde(default)]
    pub has_tools_md: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SshHostCandidate {
    pub host_alias: String,
    pub hostname: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(default = "default_ssh_port")]
    pub port: u16,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub identity_file: Option<String>,
    #[serde(default)]
    pub gateway_responds: bool,
}

fn default_ssh_port() -> u16 {
    22
}

fn default_theme() -> String {
    "system".to_string()
}

fn default_font_size() -> u32 {
    13
}

fn default_sync_interval() -> u32 {
    30
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Settings {
    #[serde(default = "default_install_method")]
    pub install_method: String,
    #[serde(default = "default_theme")]
    pub theme: String,
    #[serde(default = "default_font_size")]
    pub editor_font_size: u32,
    #[serde(default = "default_sync_interval")]
    pub sync_interval_seconds: u32,
    #[serde(default)]
    pub openclaw_connections: Vec<OpenClawConnection>,
}

fn default_install_method() -> String {
    "copy".to_string()
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            install_method: default_install_method(),
            theme: default_theme(),
            editor_font_size: default_font_size(),
            sync_interval_seconds: default_sync_interval(),
            openclaw_connections: Vec::new(),
        }
    }
}
